[Chunk Data Pack Pruner] Add Engine for pruning chunk data pack

zhangchiqing opened the pull request:
This PR adds the engine for chunk data pack pruning. By default the chunk data pack pruning is disabled, unless flag to enable it is specified.

peterargue reviewed (commented):
Thanks for creating this feature!

overall I think the approach is sound, however I find it complex with the multiple layers of wrappers and interfaces. reading it through, I've had a hard time convincing myself it's correct because there are a lot of moving pieces to keep in context. I worry that this will ballon as we add more data types to be pruned.

Is there a way to simplify this? to me it seems like there are 3 parts:

1. The pruning logic that determine which blocks to prune when
2. The height/view tracking logic that provides the next blockID to prune
3. Individual data type logic that removes all of the data and indexes associated with a blockID.

Could this be broken into 3-4 modules, plus some data type specific logic that lives with the storage module? I think having the logic in fewer places, with fewer levels of abstraction would help with maintainability and review.

zhangchiqing commented:
I designed the implementation in a more abstracted manner, allowing key logic to be reused when developing other pruners. Some important aspects include:
- Handling interruptions: Ensures the process can resume after restarts.
- Ensuring full iteration: Every block from the root to the latest is guaranteed to be iterated at least once.
- Managing a moving latest block: Since latest is dynamic, the iteration logic accounts for this.

Added Flexibility:
- The executor (`module/block_iterator/executor`) and block iterator are decoupled, preventing the block iterator from being tightly coupled with storage.
- Because the block iterator is storage-agnostic, it can be used for tasks beyond pruning, such as migrating data between databases (e.g., from Badger to Pebble).
- The existing executor is not concurrency-safe, as the pruner doesn’t require it. However, a concurrent executor can be implemented separately while still leveraging the block iterator. This is possible because the block iterator does not persist progress automatically—instead, it leaves this responsibility to the caller (executor).

This structure can support building the following modules when reusing the block_iterator and creators:
- Verification node’s approval pruner
- Execution node’s execution data pruner
- Collection node’s protocol data pruner

Regarding your question, @peterargue, I’ve added some comments—let me know if they provide the clarity you need.

janezpodhostnik commented on engine/execution/pruner/core.go:
```diff
@@ -0,0 +1,176 @@
+	}
+
+	sleeper := func() {
+		time.Sleep(config.SleepAfterEachBatchCommit)
```
this function also needs to pay attention to the context.

janezpodhostnik replied:
It might be better to just put this logic inside `IterateExecuteAndCommitInBatch`

zhangchiqing replied:
I changed the sleeper function to listen to the context.Done(). 

>It might be better to just put this logic inside IterateExecuteAndCommitInBatch

I can't, because when `time.Sleep()` starts, it can't be stopped, unless it's in a `select` statement. So I had to put the select context.Done() into this sleeper function.

peterargue reviewed (approved):
added a few small questions and comments, but otherwise this looks good

zhangchiqing commented on module/block_iterator/executor/executor.go:
```diff
@@ -46,34 +53,59 @@ func IterateExecuteAndCommitInBatch(
+		case <-ctx.Done():
+			if iteratedCountInCurrentBatch > 0 {
+				// commit the last batch
+				err := commitAndCheckpoint(log, batch, iter)
+				if err != nil {
+					return err
+				}
```
It's different, when `err == nil` we need to run `break` to go back to the loop, instead of return and exiting the loop
